Assets/Scripts/Game/EnemySenses.cs

```diff
@@ -923,24 +923,41 @@ bool CanSeeTarget(DaggerfallEntityBehaviour target)
             return seen;
         }
 
+        private static int defaultLayerOnlyMask = 0;
+        private static RaycastHit[] hitsBuffer = new RaycastHit[4];
+
         bool CanHearTarget()
         {
             float hearingScale = 1f;
 
-            // If something is between enemy and target then return false (was reduce hearingScale by half), to minimize
-            // enemies walking against walls.
-            // Hearing is not impeded by doors or other non-static objects
-            RaycastHit hit;
-            Ray ray = new Ray(transform.position, directionToTarget);
-            if (Physics.Raycast(ray, out hit))
+            if (defaultLayerOnlyMask == 0)
+                defaultLayerOnlyMask = 1 << LayerMask.NameToLayer("Default");
+
+            // TODO: Modify this by how much noise the target is making
+            if (distanceToTarget < (HearingRadius * hearingScale) + mobile.Enemy.HearingModifier)
             {
-                //DaggerfallEntityBehaviour entity = hit.transform.gameObject.GetComponent<DaggerfallEntityBehaviour>();
-                if (GameObjectHelper.IsStaticGeometry(hit.transform.gameObject))
-                    return false;
+                // If something is between enemy and target then return false (was reduce hearingScale by half), to minimize
+                // enemies walking against walls.
+                // Hearing is not impeded by doors or other non-static objects
+                Ray ray = new Ray(transform.position, directionToTarget);
+                int nhits;
+                while (true) {
+                    nhits = Physics.RaycastNonAlloc(ray, hitsBuffer, distanceToTarget, defaultLayerOnlyMask);
```

Comment:
I have issues with this approach. 

We use a fixed buffer for the multi-cast (cast with multiple results). And we're aware that the results may be partial due to buffer size. If we manage to fill the whole buffer, we retry with a larger buffer, in case we missed results.

A few things jump out to me
1) Initial buffer size is 4, which seems small
2) Buffer size increases by 1 every iteration

If a situation occurs where we hit 1000 objects at once (when entering a dungeon for the first time), we have to go through this loop up to 996 times.
I'd be much easier with an initial size of 8 (seems like it leaves room for 3-4 static models and a few entities), and a growth strategy of x2 every iteration. This would greatly limit the potential hitches due to a sudden large growth, while keeping the benefits of amortizing to "no allocation" eventually.

Reply:
Admittedly, I could not hit a case of this many objects. Mostly because the hearing radius is so small. But I'd still request the x2 growth, just in case a mod goes crazy with hearing distances and object density

Reply:
Yes, max distance is limited, but I can double array size to get O(1) amortized allocation, no problem.

By the way, there's another reason why enemies often face walls, chasing one another thru walls: stealthiness, or the lack of thereof, is implemented by creatures having some probability to periodically "leak" their current position. So it's just Daggerfall mechanics at play (exactly what [DunnyOfPenwick was saying](https://forums.dfworkshop.net/viewtopic.php?p=68078#p68078)).
So this patch will only fix the bug of non-static first obstacle allowing to see thru walls behind them.